homeassistant/components/radarr/calendar.py

```diff
@@ -0,0 +1,63 @@
+"""Support for Radarr calendar items."""
+from __future__ import annotations
+
+from datetime import datetime
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import EntityDescription
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import RadarrEntity
+from .const import DOMAIN
+from .coordinator import CalendarUpdateCoordinator, RadarrEvent
+
+CALENDAR_TYPE = EntityDescription(
+    key="calendar",
+    name=None,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Radarr calendar entity."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]["calendar"]
+    async_add_entities([RadarrCalendarEntity(coordinator, CALENDAR_TYPE)])
+
+
+class RadarrCalendarEntity(RadarrEntity, CalendarEntity):
+    """A Radarr calendar entity."""
+
+    coordinator: CalendarUpdateCoordinator
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        if not self.coordinator.event:
+            return None
+        return CalendarEvent(
+            summary=self.coordinator.event.summary,
+            start=self.coordinator.event.start,
+            end=self.coordinator.event.end,
+            description=self.coordinator.event.description,
+        )
+
+    # pylint: disable-next=hass-return-type
+    async def async_get_events(  # type: ignore[override]
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[RadarrEvent]:
+        """Get all events in a specific time frame."""
+        return await self.coordinator.async_get_events(start_date, end_date)
```

Comment:
We're not allowed to send `CalendarEvents` with additional attributes currently. We should improve the parsing code in the calendar integration to access the allowed attributes explicitly so we can remove this limitation.

CC @allenporter

Reply:
You're saying we should prevent extra attributes on the data class from being returned in the API? (E.g. extra fields added when subclassed). This didn't occur to me. Makes sense to me, we can only return fields from a fixed list.

Reply:
Yes, exactly. 👍